### ISSUE TITLE
Unknown import symbol (behave @when decorator)

### ISSUE THREAD

**Issue by werner-lehmann** [microsoft/pyright#463]
**Describe the bug**
The BDD library Behave exports decorators `given`, `when`, `then`. Those are not recognized by pyright:
![Screenshot_20200102_004605](https://user-images.githubusercontent.com/26433648/71647605-a9794f80-2cf9-11ea-83ac-ff1c4f6c8125.png)

Behave actually exports these via `__all__`, see in Github [here](https://github.com/behave/behave/blob/master/behave/__init__.py) or the code snippet below. Not affected are other imported symbols, e.g. `register_type`. The difference seems to be that the latter is explictly imported in `__int.py__`:
```python
from __future__ import absolute_import
from behave.step_registry import *      # pylint: disable=wildcard-import
from behave.matchers import use_step_matcher, step_matcher, register_type
from behave.fixture import fixture, use_fixture
from behave.version import VERSION as __version__

# pylint: disable=undefined-all-variable
__all__ = [
    "given", "when", "then", "step", "use_step_matcher", "register_type",
    "Given", "When", "Then", "Step",
    "fixture", "use_fixture",
    # -- DEPRECATING:
    "step_matcher"
]
```
Actually, pylint has the same problem and I have been silencing it as a workaround, as can be seen in the screenshot. This was reported with [this](https://github.com/behave/behave/issues/641) issue on the behave project (not by me) but turned down because the fix needed to be in pylint. According to the pyright configuration file documentation there is no option to suppress "unknown import symbol". Not sure if any of the ```reportUnknown*``` includes this as well.

**VS Code extension or command-line**
VS Code 1.41.1 on Linux
Python Extension 2019.11.50794
Pyright 1.1.16

**Additional context**
Here is an example how this is used:
```python
@then('the file "{path}" exists')
def validate_file_exists(context, path):
    assert Path(path).exists()
```

**Comment by erictraut** [comment]
The problem is that the library creates the decorators dynamically. See the bottom of [this file](https://github.com/behave/behave/blob/master/behave/step_registry.py) for the actual implementation. Static type analyzers will not be able to reason about dynamically-created symbols.

You will need to find (or create) a set of type stubs for the library. Pyright is designed for use with type stubs. It makes a best-case attempt to extract type information from the implementation (".py") files, but there are limits to what it can do given the dynamic and non-standard behaviors of many libraries. Type stubs are especially important for decorator functions.

The best solution is to have the library publisher provide type stub files (either packaged with their library or as a separate package). If the library author is unwilling to do this, you can create your own. You can use Pyright to create a "first draft" of a type stub, but you will need to do additional work to complete it.

**Comment by werner-lehmann** [comment]
Understood. After some reading, more trial and even more error I came up with these two files:

In `typings/behave/__init.pyi__`:
```python
from typings import Any, Callable, TypeVar

_FuncT = TypeVar('_FuncT', bound=Callable[..., Any])
_VT = TypeVar('_VT')

def register_type(**kwargs: _VT) -> None: ...
def given(step_text: str) -> Callable[[_FuncT], _FuncT]: ...
def when(step_text: str) -> Callable[[_FuncT], _FuncT]: ...
def then(step_text: str) -> Callable[[_FuncT], _FuncT]: ...
```

In `typings/behave/runner.pyi`:
```python
from typings import Set, Any
from behave.model import Table

class Context(object):
    table: Table
    tags: Set[str]
    def __getattr__(self, attr) -> Any: ...
    ...
```

It seems to do the trick (of silencing the errors). The `__getattr__` allows to add any custom attribute on the `Context` class which seems to be the endorsed way to do this with Behave. Since this is my first encounter to Python typing (still being new on Python itself), does this look ok? It is of course far from complete. If so I would probably create  a PR for Behave.

**Comment by erictraut** [comment]
Yeah, looks good!

One minor thing... the extra ellipsis ("...") is unnecessary in the `Context` class definition. That's needed only as a placeholder if you don't have any other variable or method declarations for the class.
